tools/wptrunner/wptrunner/testdriver-extra.js

```diff
@@ -1,25 +1,52 @@
 "use strict";
 
 (function(){
-    let pending_resolve = null;
-    let pending_reject = null;
-    window.addEventListener("message", function(event) {
-        const data = event.data;
+    let previous = Promise.resolve();
+    /**
+     * Schedule an asynchronous operation (described by a Promise-returning
+     * function) to be executed following completion of every asynchronous
+     * operation that has previously been scheduled in this way.
+     *
+     * @param {Function} async_op - Promise-returning function
+     * @returns {Promise} resolved with the result of invoking the provided
+     *                    function
+     */
+    function enqueue(async_op) {
+        const next = previous.then(async_op);
+        previous = next.catch(function() {});
```

Comment:
Should we add a failure message here saying where we were in the chain? Might be useful to debugging, but also might add needless complexity.

Reply:
We're returning `next` to the caller, so they still have the ability to react to errors.

That said, this pattern *does* hide unhandled rejections. That's because even though tests still have the option to react to failures in the `next` promise, the invocation of `catch` means that there will always be at least one handler.

Here's a more involved solution that avoids this:

```js
function enqueue(async_op) {
    let next;
    previous = new Promise(function(resolve) {
        next = previous.then(async_op)
          .then(resolve, function(err) {
              resolve();
              throw err;
          });
    });
    return next;
}
```

This feels a little too clever to me, but for a small and well-defined utility function like `enqueue`, it might be the right thing to do. What do you think?